staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc.go

```diff
@@ -137,159 +210,123 @@ func New(opts OIDCOptions) (*OIDCAuthenticator, error) {
+		usernameClaim:  opts.UsernameClaim,
+		usernamePrefix: opts.UsernamePrefix,
+		groupsClaim:    opts.GroupsClaim,
+		groupsPrefix:   opts.GroupsPrefix,
+		ctx:            ctx,
+		cancel:         cancel,
 	}
-	return
-}
 
-func (a *OIDCAuthenticator) client() (*oidc.Client, error) {
-	// Fast check to see if client has already been initialized.
-	if client := a.oidcClient.Load(); client != nil {
-		return client.(*oidc.Client), nil
+	now := opts.now
+	if now == nil {
+		now = time.Now
 	}
 
-	// Acquire lock, then recheck initialization.
-	a.mu.Lock()
-	defer a.mu.Unlock()
-	if client := a.oidcClient.Load(); client != nil {
-		return client.(*oidc.Client), nil
+	verifierConfig := &oidc.Config{
+		ClientID:             opts.ClientID,
+		SupportedSigningAlgs: supportedSigningAlgs,
+		Now:                  now,
 	}
 
-	// Try to initialize client.
-	providerConfig, err := oidc.FetchProviderConfig(a.httpClient, a.issuerURL)
-	if err != nil {
-		glog.Errorf("oidc authenticator: failed to fetch provider discovery data: %v", err)
-		return nil, fmt.Errorf("fetch provider config: %v", err)
-	}
+	initVerifier(authenticator, verifierConfig)
+	return authenticator, nil
+}
 
-	clientConfig := oidc.ClientConfig{
-		HTTPClient:     a.httpClient,
-		Credentials:    oidc.ClientCredentials{ID: a.trustedClientID},
-		ProviderConfig: providerConfig,
+func hasCorrectIssuer(iss, tokenData string) bool {
+	parts := strings.Split(tokenData, ".")
+	if len(parts) != 3 {
+		return false
 	}
-
-	client, err := oidc.NewClient(clientConfig)
+	payload, err := base64.RawURLEncoding.DecodeString(parts[1])
 	if err != nil {
-		glog.Errorf("oidc authenticator: failed to create client: %v", err)
-		return nil, fmt.Errorf("create client: %v", err)
+		return false
 	}
-
-	// SyncProviderConfig will start a goroutine to periodically synchronize the provider config.
-	// The synchronization interval is set by the expiration length of the config, and has a minimum
-	// and maximum threshold.
-	stop := client.SyncProviderConfig(a.issuerURL)
-	a.oidcClient.Store(client)
-	a.close = func() {
-		// This assumes the stop is an unbuffered channel.
-		// So instead of closing the channel, we send am empty struct here.
-		// This guarantees that when this function returns, there is no flying requests,
-		// because a send to an unbuffered channel happens after the receive from the channel.
-		stop <- struct{}{}
+	claims := struct {
+		// WARNING: this JWT is not verified. Do not trust these claims.
+		Issuer string `json:"iss"`
+	}{}
+	if err := json.Unmarshal(payload, &claims); err != nil {
+		return false
 	}
-	return client, nil
+	if claims.Issuer != iss {
+		// Google sometimes returns "accounts.google.com" instead of "https://accounts.google.com"
+		// in its ID tokens. Allow this case.
+		//
+		// https://developers.google.com/identity/protocols/OpenIDConnect#obtainuserinfo
+		if iss == "https://accounts.google.com" && claims.Issuer == "accounts.google.com" {
```

Comment:
I don't think we should embed vendor-specific non-spec behavior like this. It's hard to justify writing to the spec (e.g. our position on use of email_verified) when combined with exceptions. @mikedanese, do you know under what circumstances the issuer does not match the published value in https://accounts.google.com/.well-known/openid-configuration?

Reply:
go-oidc has always had this behavior to support Google. It'd be a breaking change to merge this change without this, though I agree with the sentiment.

Reply:
Background is we implemented this while the spec was still under development and the fully qualified https:// prefixed uris was a late addition. We serve two token endpoints, on v3/ and v4/ which issue accounts.google.com and https://accounts.google.com issued id tokens respectively. The openid-configuration currently points to v4/.

Reply:
Background is we implemented this while the spec was still under development and the fully qualified https:// prefixed uris was a late addition. We serve two token endpoints, on v3/ and v4/ which issue accounts.google.com and https://accounts.google.com issued id tokens respectively. The openid-configuration currently points to v4/.

Reply:
@mdietz is source of truth on this

Reply:
The issuer returned in a Google ID Token depends on the token endpoint used.  The accounts.google.com and googleapis.com/oauth2/v3 token endpoints have a pre-finalized OIDC spec implementation that elides the https:// in the issuer field for issued ID tokens.  The googleapis.com/oauth2/v4 token endpoint is spec compliant.  See http://openid.net/specs/openid-connect-core-1_0.html#GoogleIss.

Reply:
Sounds like if we use the URLs in the discovery doc, which we do, we should be able to rely on correct issuers now.

Reply:
@ericchiang ping on removing the google issuer spec exception, since the published token endpoint is now spec compliant

Reply:
Sorry wanted to verify this manually. Tested and can confirm @mdietz's comments.